src/ucp/rndv/proto_rndv.c

```diff
@@ -365,41 +348,122 @@ ucp_proto_rndv_ctrl_init(const ucp_proto_rndv_ctrl_init_params_t *params,
+                                  void *priv, ucp_proto_config_t *cfg)
+{
+    cfg->priv           = priv;
+    cfg->ep_cfg_index   = init_params->ep_cfg_index;
+    cfg->proto          = ucp_protocols[proto->proto_id];
+    cfg->rkey_cfg_index = init_params->rkey_cfg_index;
+    cfg->select_param   = *select_param;
 }
 
-void ucp_proto_rndv_ctrl_probe(const ucp_proto_rndv_ctrl_init_params_t *params)
+void ucp_proto_rndv_ctrl_probe(const ucp_proto_rndv_ctrl_init_params_t *params,
+                               void *priv, size_t priv_size)
 {
-    ucp_proto_rndv_ctrl_priv_t priv;
+    const ucp_proto_init_params_t *init_params = &params->super.super;
+    ucp_proto_rndv_ctrl_priv_t *rpriv          = priv;
+    const ucp_proto_select_init_protocols_t *remote_proto_init;
+    ucp_proto_perf_range_t ctrl_perf, *last_range;
+    ucp_proto_select_param_t remote_select_param;
+    ucp_proto_select_elem_t *remote_proto_select;
+    ucp_proto_init_elem_t *remote_proto;
+    ucp_proto_caps_t *remote_caps;
     ucp_proto_caps_t caps;
     ucs_status_t status;
+    size_t min_length, max_length, cfg_thresh;
+    void *remote_priv;
+
+    ucs_assert(params->super.flags & UCP_PROTO_COMMON_INIT_FLAG_RESPONSE);
+    ucs_assert(!(params->super.flags & UCP_PROTO_COMMON_INIT_FLAG_SINGLE_FRAG));
 
-    status = ucp_proto_rndv_ctrl_init(params, &caps, &priv);
+    if (!ucp_proto_common_init_check_err_handling(&params->super)) {
+        return;
+    }
+
+    /* Initialize 'rpriv' structure */
+    ucp_proto_rndv_ctrl_init_priv(params, rpriv, params->lane);
+
+    /* Init remote proto */
+    remote_select_param = ucp_proto_rndv_remote_select_param_init(params);
+    status              = ucp_proto_rndv_ctrl_select_remote_proto(
+            params, &remote_select_param, rpriv->md_map, &remote_proto_select);
     if (status != UCS_OK) {
         return;
     }
+    remote_proto_init = &remote_proto_select->proto_init;
+
+    /* Add variants for each remote proto */
+    ucs_array_for_each(remote_proto, &remote_proto_init->protocols) {
+        remote_caps = &remote_proto->caps;
+        /* Skip empty variants and reconfig proto */
+        if ((remote_caps->num_ranges == 0) || (remote_proto->proto_id == 0)) {
+            continue;
+        }
+
+        last_range = remote_caps->ranges + (remote_caps->num_ranges - 1);
+        max_length = ucs_min(params->super.max_length, last_range->max_length);
+        min_length = ucs_max(params->super.min_length,
+                             remote_proto->caps.min_length);
+        if (max_length < min_length) {
+            continue;
+        }
+
+        remote_priv = &ucs_array_elem(&remote_proto_init->priv_buf,
+                                      remote_proto->priv_offset);
+        ucp_proto_rndv_set_variant_config(init_params, remote_proto,
+                                          &remote_select_param, remote_priv,
+                                          &rpriv->remote_proto_config);
```

Comment:
That's not seem to be possible since we also need to obtain pointer to `priv` buffer, `ep_cfg_index` and `remote_select_param` to call remote protocol query during RTS/RTR query call.

We can store these variables separately, but from my point of view it's simpler and more straightforward to store the whole `ucp_proto_config_t` structure.

Reply:
ok